File: mari-server/src/models/Transaction.js
// src/models/Transaction.js
const mongoose = require('mongoose');

const transactionSchema = new mongoose.Schema({
  transactionId: {
    type: String,
    required: true,
    unique: true
  },
  senderBioHash: {
    type: String,
    required: true
  },
  receiverBioHash: {
    type: String,
    required: true
  },
  amount: {
    type: Number,
    required: true,
    min: 0
  },
  locationGrid: {
    type: String,
    required: true
  },
  timestamp: {
    type: Date,
    default: Date.now
  },
  status: {
    type: String,
    enum: ['PENDING', 'SETTLED', 'FAILED'],
    default: 'PENDING'
  },
  coupon: {
    type: String,
    required: true
  },
  transportMethod: {
    type: String,
    enum: ['SMS', 'HTTP'],
    required: true
  },
  physicsData: {
    location: {
      latitude: Number,
      longitude: Number,
      grid: String
    },
    motion: {
      x: Number,
      y: Number,
      z: Number
    },
    timestamp: Date
  }
}, {
  timestamps: true
});

transactionSchema.index({ transactionId: 1 });
transactionSchema.index({ senderBioHash: 1 });
transactionSchema.index({ receiverBioHash: 1 });
transactionSchema.index({ status: 1 });
transactionSchema.index({ createdAt: -1 });

module.exports = mongoose.model('Transaction', transactionSchema);

File: mari-app/domain/model/Transaction.kt
package com.Mari.mobileapp.domain.model

enum class TransactionStatus {
    PENDING, COMPLETED, FAILED
}

enum class TransactionType {
    SEND, RECEIVE
}

enum class TransportMethod {
    SMS, RECEIVED
}

/**
 * Transaction domain model
 * 
 * IMPORTANT: "senderBioHash" and "receiverBioHash" are TERRIBLE variable names!
 * They are NOT biometric data - they're just pseudonymous user account identifiers.
 * Think of them as: senderId, receiverId, senderAccountId, receiverAccountId
 * 
 * NO special biometric hardware is needed - these are just random unique IDs
 * that identify users in the system (like account numbers).
 */
data class Transaction(
    val id: String,
    
    // WARNING: Misleading names! These are NOT biometric data.
    // They're just pseudonymous user identifiers (like account IDs).
    // Should be renamed to: senderId, receiverId, etc.
    val senderBioHash: String,      // TODO: Rename to senderId
    val receiverBioHash: String,    // TODO: Rename to receiverId
    
    val amount: Double,
    val locationGrid: String,
    val timestamp: Long,
    val status: TransactionStatus,
    val type: TransactionType,
    val coupon: String,
    val transportMethod: TransportMethod
)
